nemo/collections/asr/modules/rnnt.py

```diff
@@ -50,6 +50,525 @@
 from nemo.utils import logging
 
 
+class StatelessNet(torch.nn.Module):
+    def __init__(self, context_size, vocab_size, emb_dim, blank_idx, blank_as_pad, normalization_mode, dropout):
+        super().__init__()
+        self.context_size = context_size
+        self.vocab_size = vocab_size
+        self.emb_dim = emb_dim
+        self.dropout = torch.nn.Dropout(dropout)
+        self.norm = torch.nn.Identity()
+        if normalization_mode == 'layer':
+            self.norm = torch.nn.LayerNorm(emb_dim, elementwise_affine=False)
+
+        assert(blank_as_pad)  # So far it only supports when it is set True
+        embeds = []
+        for i in range(self.context_size):
+            # We use different embedding matrices for different context positions.
+            # In this list, a smaller index means more recent history word.
+            # We assign more dimensions for the most recent word in the history.
+            if i != 0:
+                embed_size = emb_dim // 2 // self.context_size
+            else:
+                embed_size = emb_dim - (emb_dim // 2 // self.context_size) * (self.context_size - 1)
+
+            if blank_as_pad:
+                embed = torch.nn.Embedding(vocab_size + 1, embed_size, padding_idx=blank_idx)
+            else:
+                embed = torch.nn.Embedding(vocab_size, embed_size)
+
+            embeds.append(embed)
+
+        self.embeds = torch.nn.ModuleList(embeds)
+        self.blank_idx = blank_idx
+
+    def forward(self,
+                y: Optional[torch.Tensor] = None,
+                state: Optional[List[torch.Tensor]] = None,
+                ):
+        # Although this is a *stateless* net, we use the "state" parameter to
+        # pass in the previous labels, unlike LSTMs where state would represent
+        # hidden activations of the network.
+        # "state" is a list of 1 tensor in order to be consistent with the stateful
+        # decoder interface, and the element is a tensor of shape [batch, context-length].
+        # The out dimension of this function is B x U x D, with D being the total embedding dim.
+        outs = []
+
+        [B, U] = y.shape
+        appended_y = y
+        if state != None:
+            appended_y = torch.concat([state[0], y], axis=1)
+            context_size = appended_y.shape[1]
+
+            if context_size < self.context_size:
```

Comment:
Add a comment stating which codepath has how much future context (from first glance, it seems to be amount of "self.context_size".

Reply:
Done